### PR TITLE
AMM-CDA-9 Adds soft and hard failure distinction for AMM and order book errors

### PR DESCRIPTION
**Pull request by Chralt98**

### CODE REVIEW

File: zrml/neo-swaps/src/lib.rs [hunk 1]
```diff
@@ -1005,6 +1005,26 @@ mod pallet {
...
+                Error::<T>::NumericalLimits(NumericalLimitsError::MinAmountNotMet).into();
+            if spot_price_too_low == error
+                || spot_price_slipped_too_low == error
+                || max_amount_exceeded == error
+                || min_amount_not_met == error
+            {
+                ApiError::SoftFailure(AmmSoftFail::Numerical)
+            } else {
+                ApiError::HardFailure(error)
+            }
```
Comment (maltekliemann) [inline zrml/neo-swaps/src/lib.rs:1026]: I'd make this a match statement.

File: zrml/orderbook/src/lib.rs [hunk 2]
```diff
@@ -508,6 +508,21 @@ mod pallet {
         }
     }
 
+    impl<T: Config> Pallet<T> {
+        fn match_failure(error: DispatchError) -> ApiError<OrderbookSoftFail> {
+            let below_minimum_balance: DispatchError = Error::<T>::BelowMinimumBalance.into();
+            let partial_fill_near_full_fill_not_allowed: DispatchError =
+                Error::<T>::PartialFillNearFullFillNotAllowed.into();
+            if error == below_minimum_balance {
```
Comment (maltekliemann) [inline zrml/orderbook/src/lib.rs:516]: I'd make this a match statement.

**Review by maltekliemann — APPROVED** [review]

File: zrml/hybrid-router/Cargo.toml [hunk 3]
```diff
@@ -57,6 +58,7 @@ mock = [
     "zrml-prediction-markets/default",
     "zrml-prediction-markets/mock",
     "zrml-simple-disputes/default",
+    "zrml-swaps/default",
```
[the remark this answers is not in the document]
Comment (Chralt98) [reply zrml/hybrid-router/Cargo.toml]: e89f24894103a8e1d86ce4887b0b701ded3be092

File: zrml/neo-swaps/src/lib.rs (hunk 1 above)
Comment (Chralt98) [reply zrml/neo-swaps/src/lib.rs:1026]: I wanted this so bad too.. But I found no working way to do it. I would be pleased if you can make a suggestion that works :-) Thanks a lot!

Comment (maltekliemann) [reply zrml/neo-swaps/src/lib.rs:1026]: Ah yeah, nevermind. I thought this was possible, but apparently I don't understand how match statements work.

**Review by maltekliemann — APPROVED** [review]

**Pull request merged**